### ISSUE TITLE
Keycloak Organizations

### ISSUE THREAD

**Issue by pedroigor** [keycloak/keycloak#30180]
### Description

This top-level epic is about the planned work to deliver the first release of the Keycloak Organization feature.

Keycloak Organization is a feature that leverages and builds on top of the Identity and Access Management (IAM) capabilities of Keycloak to address Customer Identity and Access Management (CIAM) with a focus on Business-to-Business (B2B) use cases.

                                        +-----------+                  
                                        |           |                  
                            +-----------+  Customer +-----------+      
                            |           |           |           |      
    +----------+      +-----+----+      |           |      +----+-----+
    |          |      |          |      +-----------+      |          |
    | End-User +------+  Realm   |                         | End-User |
    |          |      |          |                         |          |
    |          |      |          |                         |          |
    +----------+      +-----+----+                         +----+-----+
                            |           +-----------+           |      
                            |           |           |           |      
                            +-----------+  Partner  +-----------+      
                                        |           |                  
                                        |           |                  
                                        +-----------+                       

In summary, this feature will allow a realm to integrate with third-party entities like customers and business partners so that their identities and how they access protected resources at the realm level are segregated and managed within the scope of the organization they belong to. 

As a result, a realm will be able to provide a different experience when onboarding customer identities or identities from a business partner and mechanisms to secure and manage how they interact with a realm and how they access protected resources from the clients available from a realm. Although not the main focus, it should also be possible to leverage this feature if you just want a better segregation between users, clients, and how they are managed and authenticate to a realm.

The main set of capabilities provided by Keycloak Organizations are:

* * [x] **Manage Organizations**
* * [ ] **Manage Organization Members**
* * * [x] Organization Member Self-Registration based on Identity Brokering
* * * [x] Organization Member Registration based on Invitation Links
* * * [ ] Joining Multiple Organizations 
* * [x] **Mapping organization metadata in tokens** 
* * [ ] **Manage Organization Roles**
* * [ ] **Manage Organization Groups**
* * [ ] **Manage Authentication within the Scope of an Organization**
* * * [ ] Authentication based on the Organization Authentication Policies**
* * [ ] **Manage Clients and Service Accounts**
* * [ ] **Administrative Authorization**
* * * [ ] **RBAC**
* * * [ ] **Fine-Grained** 
* * [ ] **Organization Self-Service**

The planned release dates for Keycloak Organization are:

| Release  | State |
| ------------- | ------------- |
| 25.0.0  | Preview  |
| 26.0.0  | Supported  |

For more details about the release dates, see https://github.com/keycloak/keycloak/milestones.

### Discussion

https://github.com/keycloak/keycloak/discussions/23948

### Issues

- [First Release of Keycloak Organizations](https://github.com/keycloak/keycloak/issues/28609)

### Motivation

Keycloak is a well-known open-source IAM solution and we want to leverage now its capabilities to enable CIAM with a focus on B2B use cases.

**Comment by brecht-vermeersch** [comment]
Thanks for all the work! Is joining multiple organisations a confirmed feature that will be added?

**Comment by pedroigor** [comment]
@brecht-vermeersch Yes. I've proposed it here https://github.com/keycloak/keycloak/issues/30229#issuecomment-2153353571.

Although not originally planned, I think it will add more value to this first iteration as it seems to be a common request from people interested in this feature ...

**Comment by osajdapawel** [comment]
Great work guys, thank you.

I got question about Organization Roles. Do you plan to add the feature in any of upcoming versions?

I got also a suggestion about linking Identity Providers to organization. 
Lets go with example:
We got Microsoft multitenant IdProvider and now it cannot be linked to more than one provider. 
It would be interesting to have possibility to link it to more than one organization and filter user by email domain and add users to organization that matches that email address (I don't know all the staff behind so maybe it is missed suggestion)

**Comment by pedroigor** [comment]
> Great work guys, thank you.
> 
> I got question about Organization Roles. Do you plan to add the feature in any of upcoming versions?

Yes, but not for Keycloak 26. See https://github.com/keycloak/keycloak/issues/30229.

> 
> I got also a suggestion about linking Identity Providers to organization. Lets go with example: We got Microsoft multitenant IdProvider and now it cannot be linked to more than one provider. It would be interesting to have possibility to link it to more than one organization and filter user by email domain and add users to organization that matches that email address (I don't know all the staff behind so maybe it is missed suggestion)

Is it related to @osajdapawel Is it about https://github.com/keycloak/keycloak/issues/30088?

**Comment by osajdapawel** [comment]
Thank you for quick answer.

I meant that one Identity Provider linked to many organizations.
For example one global multitenant Microsoft Identity Provider (that can interact with many different domains) could be linked  to many already predefined organizations that would filter users based on domain name.

**Comment by pedroigor** [comment]
@osajdapawel Today you need to create individual MS IdPs, with the same configuration, and link them to each organization.

There is a constraint that a broker can only be linked to a single organization. But we can review this. Do you want to create an issue and see if we get some votes to support this use case?

**Comment by y-tabata** [comment]
@pedroigor 
Is there a use case where not only themes but also SPI providers are implemented (customized) for each organization?
When considering this use case, the following constraints of the Quarkus distribution concern me. This is because I think it would be difficult to manage rules such as package names across organizations.

> With the new distribution there is no longer a separate classpath for custom providers, so you may need to be more careful with what additional dependencies you include.

**Comment by UXabre** [comment]
This is awesome! This seems like the path we definitely want to migrate to. Perhaps, is there a way to skip the identification phase if we know the organization up front? Because, if I understand correctly, in the future we'll be able to link identity providers to different organizations. Will this work all the same for federations as well, in which an organization admin will be able to add an LDAP link specific for their organization?
Thank you and everyone involved for the great work!

**Comment by RyanRad-BSG** [comment]
Good evening - this is awesome. I've been searching all day trying to find out the best way to configure our realm to have internal & external groups with different client accesses. After an exhaustive search I found this new preview feature. Seems like with some time, specific client allocation to Organizations will be added? I'd like external users to have limited access to clients while granting access to all clients for internal users. I activated the preview features, doesn't look like that can be accomplished with the first release, but should come available down the line. Is my understanding correct?

**Comment by aljaznuncic** [comment]
I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.

I hope that in the future there will also be an "admin" console for administrators of individual organizations.

**Comment by osajdapawel** [comment]
> @osajdapawel Today you need to create individual MS IdPs, with the same configuration, and link them to each organization.
> 
> There is a constraint that a broker can only be linked to a single organization. But we can review this. Do you want to create an issue and see if we get some votes to support this use case?

@pedroigor we could try to make this issue (to check if one IdentityProvider can be linked to many organizations)

**Comment by Florzl** [comment]
Hi,
Does the item "Manage Clients and Service Accounts" mean something like the one suggested in my ticket, if so I will close the ticket #31164 

Thanks!

**Comment by rspu** [comment]
> Perhaps, is there a way to skip the identification phase if we know the organization up front? 

+1 for the suggestion by @UXabre. I would also like to have an option to pre-select the organization for the browser authentication flow if I already know the organization in the application initiating the authentication flow (e.g., multi-tenant web app with the organization ID in the subdomain: <organization id>.myapp.com).

**Comment by pedroigor** [comment]
Hi all,

Can I ask you to give your thoughts and help with what you expect from the support of users joining multiple organizations?

Please, add any comments here https://github.com/keycloak/keycloak/issues/30747.

**Comment by Degoya-Studio** [comment]
Hi @pedroigor are you planning on creating an admin role, that is only able to invite users to their organization?

**Comment by oculos** [comment]
Another possibility I'd like to see here if possible would be:

- multiple frontend url's. Today, realms can have only one frontend url (besides the instance hostname). It would be nice if each organization could have its own frontend url
- custom themes per organization - together with the previous idea, a user coming from a given domain (frontend url) would get the look and feel of his organization.

**Comment by NiklasA** [comment]
Hi @pedroigor, one question regarding "Support for synchronizing identities from organizations using SCIM":

Does this mean that in the future it will be possible to read users from Keycloak using the SCIM Protocol, or is this more about importing users from other servers into Keycloak?

**Comment by Degoya-Studio** [comment]
It would be also nice if we could access the organization attributes inside of the id_token, for example:

```
  "organization": {
    "myOrg": {
     "key": "value"
    }
  },
```

**Comment by osajdapawel** [comment]
Hey, Is it possible now to create organization via API using own defined client permission. 
I'm struggling with create it, I can only create with admin-cli client

**Comment by Degoya-Studio** [comment]
Hi @pedroigor another good feature would be, if users can only self register with an invitation link.

**Comment by pedroigor** [comment]
@Degoya-Studio Can you please file an issue? We are collecting feedback from issues so that we can see what we can and what makes to deliver until Keycloak 26.

**Comment by pedroigor** [comment]
@osajdapawel Right now, you only need the `manage-realm` role to have access to the Organization API. Can you please elaborate?

**Comment by pedroigor** [comment]
> Hi @pedroigor, one question regarding "Support for synchronizing identities from organizations using SCIM":
> 
> Does this mean that in the future it will be possible to read users from Keycloak using the SCIM Protocol, or is this more about importing users from other servers into Keycloak?

I would say in both ways. There are different applications for SCIM where Keycloak can be the client and the server.

**Comment by pedroigor** [comment]
> Another possibility I'd like to see here if possible would be:
> 
> * multiple frontend url's. Today, realms can have only one frontend url (besides the instance hostname). It would be nice if each organization could have its own frontend URL
> * custom themes per organization - together with the previous idea, a user coming from a given domain (frontend url) would get the look and feel of his organization.

This one is a frequent demand from the community and not specifically to organizations. But usually related to multi-tenancy. I'm not sure if we'll be able to deliver it in Keycloak 26 but please, create an issue so we can track the requirement.

Right now our hostname provider and other key parts of the server are based on the idea of a single tenant.

> * the identity-first feature could then harness the frontend url. if you come from a given domain, that's your organization. No need to enter the e-mail first on the login.

I'm working on https://github.com/keycloak/keycloak/issues/31438. Which is slightly related to that but without the frontend url.

> 
> Of course, this all could be configurable, so that the existing flow could still be used instead of this.

**Comment by pedroigor** [comment]
> Hi @pedroigor are you planning on creating an admin role, that is only able to invite users to their organization?

Can you please create an issue? It makes sense for me.

**Comment by pedroigor** [comment]
> Hi, Does the item "Manage Clients and Service Accounts" mean something like the one suggested in my ticket, if so I will close the ticket #31164
> 
> Thanks!

I think so. The `Manage Clients and Service Accounts` is about making it possible to manage clients and SAs within an organization. You can have many of them you want just like you have in the realm but they are scoped to the organization.

**Comment by pedroigor** [comment]
> > Perhaps, is there a way to skip the identification phase if we know the organization up front?
> 
> +1 for the suggestion by @UXabre. I would also like to have an option to pre-select the organization for the browser authentication flow if I already know the organization in the application initiating the authentication flow (e.g., multi-tenant web app with the organization ID in the subdomain: .myapp.com).

@rspu Does it help https://github.com/keycloak/keycloak/issues/31438?

In regards to themes, if you try a nightly image you'll be able to access the organization that matched the user trying to authenticate from the `org` bean. So you can customize the pages accordingly to the current organization.

**Comment by pedroigor** [comment]
> I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.
> 
> I hope that in the future there will also be an "admin" console for administrators of individual organizations.

@aljaznuncic The "self-service" for organization administrations in among the capabilities planned for this feature. See the description of the issue. But we are not delivering it now in Keycloak 26.

In regard to the domain, does it work if you provide some random/dummy domain? Could you please elaborate more about your use case and if just the fact that the user is a member of the org is enough for you as well as being able to introspect tokens to check the organizations the user is a member?

**Comment by pedroigor** [comment]
> @pedroigor Is there a use case where not only themes but also SPI providers are implemented (customized) for each organization? When considering this use case, the following constraints of the Quarkus distribution concern me. This is because I think it would be difficult to manage rules such as package names across organizations.
> 
> > With the new distribution there is no longer a separate classpath for custom providers, so you may need to be more careful with what additional dependencies you include.

@y-tabata True. But it will depend on the packages you are using for each extension so that they won't collide. In regards to themes, I guess it will be the same as long as the names of the themes don't collide.

Or do you have something else in mind?

**Comment by pedroigor** [comment]
> Good evening - this is awesome. I've been searching all day trying to find out the best way to configure our realm to have internal & external groups with different client accesses. After an exhaustive search I found this new preview feature. Seems like with some time, specific client allocation to Organizations will be added? I'd like external users to have limited access to clients while granting access to all clients for internal users. I activated the preview features, doesn't look like that can be accomplished with the first release, but should come available down the line. Is my understanding correct?

@RyanRad-BSG Yes, that is correct. The support for clients (and SAs), as well as groups, roles, etc, will not be delivered in this first iteration. This issue should help all of us to track the different capabilities and which ones we should prioritize over others depending on the feedback.

**Comment by pedroigor** [comment]
> This is awesome! This seems like the path we definitely want to migrate to. Perhaps, is there a way to skip the identification phase if we know the organization up front? Because, if I understand correctly, in the future we'll be able to link identity providers to different organizations. Will this work all the same for federations as well, in which an organization admin will be able to add an LDAP link specific for their organization? Thank you and everyone involved for the great work!

@UXabre In regard to skipping the identification phase, it seems related to https://github.com/keycloak/keycloak/issues/31438.

As per LDAP, we don't have any requirement around this yet. But please, create an issue so we can track.

**Comment by Florzl** [comment]
> > Hi, Does the item "Manage Clients and Service Accounts" mean something like the one suggested in my ticket, if so I will close the ticket #31164
> > Thanks!
> 
> I think so. The `Manage Clients and Service Accounts` is about making it possible to manage clients and SAs within an organization. You can have many of them you want just like you have in the realm but they are scoped to the organization.

Thanks for your reply, thats very cool, is this also coming in the v26?

**Comment by rogerioefonseca** [comment]
That sounds fantastic! 
Do you have an idea of which release you’re planning to include all 21 tasks listed in the issue? We're currently leaning towards using multiple Realms, but one Realm with multiple organizations would fit much better.

**Comment by kristiyan2022** [comment]
Looking forward to this! 

Hi @pedroigor, we have a use case where the same user can belong to multiple organizations, each with different roles. When the account scope is requested, if the user holds the accountable role in a particular organization, we will attach a custom claim to the token introspection using a token mapper. For example:
| Organisation   | Realm Role       | Claim                       |
|---------------|----------------|--------------------|
| X                      | `user`               |                                  |
| Y                      | `accountable`  | account:  VALUE    |

Currently, we have assigned the `accountable` role to the list of roles in the `account` client scope. We also have a custom protocol mapper that will map the claim if the user is indeed accountable in the organisation (to prevent that the same user authenticating through organisation **X** getting the claim).

I hope you can see the complexity of this also in other contexts different than ours. My question is, is there a plan to apply only the roles the user has in the specific organisation he is authenticating from?

Thank you!

**Comment by ggomarighetti** [comment]
Hi @pedroigor!

I am captivated by the new motivation and the new direction keycloak is taking with this "it is a well-known open source IAM solution and we now want to leverage its capabilities to enable CIAM with a focus on B2B use cases".

I was following the discussion thread as well as the different pull requests that were approved, but something caught my attention from the last one (https://github.com/keycloak/keycloak/pull/29613). Given this new feature and reading the roadmap that is published in this discussion, it would seem interesting to add:

- to be able to place rules or configurations, if a user can belong or not to more than one organization.
- in case the user belongs to more than one organization, to be able to modify the browser flow to show the different options and choose one with which to enter.
- eliminate or make optional the "domain" option in the organizations, since not always customers have a corporate email.

And based on the tasks that are already planned but not yet developed, I do not know if the topic is still under discussion or the debate has not been opened yet, but, regarding authorization, I would understand the feature of "organization" as something to differentiate users, i.e., that created an organization, one can authorize it to access (or not) a certain application, and that X user in X organization can have X role, and if he/she is a member of two organizations simultaneously, he/she can have two different roles, one in each one, one can authorize it to access (or not) to a certain application, and that X user in X organization can have X role, and if he is a member of two organizations simultaneously, he can have two different roles, one in each one of them, as to follow or implement a simplification logic.

very good decision to include this in the project and this conversion to a ciam b2b tool, from already a greeting and I hope to continue debating this topic, success!

**Comment by timyourivh** [comment]
The way this is implemented as of right now looks similar to the Phase Two plugin: https://github.com/p2-inc/keycloak-orgs

**Comment by pedroigor** [comment]
@tamis-laan I would suggest you wait for Keycloak 26 to use it in production. The reason is that the feature will be fully supported in that version and we had a lot of changes and new capabilities since we released Keycloak 25.

However, I suggest you try out what we have now so that you can check how it will address your requirements and use cases. For that, consider looking at our nightly build/image as they are very close to what we are delivering in Keycloak 26.

Your feedback will be valuable so that we can try to incorporate additional capabilities before Keycloak 26. See https://github.com/keycloak/keycloak/milestones.

**Comment by tamis-laan** [comment]
@pedroigor 

Well we have our first two clients right now, so we can't really wait for v26. Would it be possible to have the users be generic users and add them to a organization later in v26? When organizations are enabled are they enforced? How will backward compatibility work?

**Comment by pedroigor** [comment]
You are able to add existing realm users to organizations.

When organizations are enabled, users will authenticate to the realm and tokens will be issued with metadata about their organization membership. Not sure what you mean by "enforced" though, can you please elaborate?

As you know, the feature is a technology preview so backward compatibility is not guaranteed. That is why I'm recommending to wait for Keycloak 26. Once the feature is supported (after 26) backward compatibility is guaranteed.

**Comment by paulwer** [comment]
Hi folks,
I wanted to share some thoughts on the rights and roles for organizations. Maybe both thoughts are goong in the same way ;)

1. We would like to see some coupleing between realm/client roles/rights and inheritance to the orgs and their members. Some kind of way to hierarchial assign roles/rights to orgs, which then can be assigned to users.

2. Organization specific roles and realm roles for orgs/org members 
within p2-inc each org had their default set of roles and when wanted a distinct role have to been created for this one organization. Keycloak would fit more use cases, when both would be possible.

**Comment by tamis-laan** [comment]
> You are able to add existing realm users to organizations.
> 
> When organizations are enabled, users will authenticate to the realm and tokens will be issued with metadata about their organization membership. Not sure what you mean by "enforced" though, can you please elaborate?
> 
> As you know, the feature is a technology preview so backward compatibility is not guaranteed. That is why I'm recommending to wait for Keycloak 26. Once the feature is supported (after 26) backward compatibility is guaranteed.

By enforce I mean that when we move to 26 from 25 (without the experimental feature) will existing users get a "default org" assigned? Or is there something like a user without any org?

**Comment by NiklasA** [comment]
> Hi @pedroigor, one question regarding "Support for synchronizing identities from organizations using SCIM":
> 
> Does this mean that in the future it will be possible to read users from Keycloak using the SCIM Protocol, or is this more about importing users from other servers into Keycloak?

Hi @pedroigor, sorry to bother you, could you please answer my question? Thanks in advance, and best regards!

**Comment by Degoya-Studio** [comment]
@pedroigor is it possible, that when I send someone an invitation to my organization and they register, that they are redirected to my web application? so basically the home url of my keycloak client

**Comment by magoogli** [comment]
Hi @pedroigor I'm not sure if it's on the road map or not, but it would be great if I could map the id of the organization ( and any organization attribute for that matter ) into the access token.  At the moment it seems only the name is added. Thank you for this awesome feature!

**Comment by codeindustriesnl** [comment]
Will the organisations-feature support different (and/or extra) "User Profile > Attributes" per organisation?

**Comment by migu83** [comment]
Did I miss something or is there currently no possibility for any organization specific configuration on a per user basis?
My use-case has been mentioned often: userA is member of orgA (has RoleX here) and orgB (has RoleY here).

I can see that a lot of effort was put into the orga feature which is great, but I could not find an outlook to when this core use-case will be covered so it is hard to actually plan using it. 26.1? 27 (in 2026?!)
Is there any plan or simple workaround I cannot see?

**Comment by Fydon** [comment]
> > Hi @pedroigor, one question regarding "Support for synchronizing identities from organizations using SCIM":
> > Does this mean that in the future it will be possible to read users from Keycloak using the SCIM Protocol, or is this more about importing users from other servers into Keycloak?
> 
> I would say in both ways. There are different applications for SCIM where Keycloak can be the client and the server.

@NiklasA There [was a response to your question](https://github.com/keycloak/keycloak/issues/30180#issuecomment-2251228776) (quoted above) but you we're mentioned.

Also interested to see what SCIM functionality will be available. I would like to synchronise account creation and deletion between two services. Currently I would need to use a service specifically designed for Keycloak rather than SCIM.

**Comment by Steve-Rithmik** [comment]
Is there any roadmap for when the additional features will be added? The organization self-service feature is a core capability that we need before we can use keycloak. Literally every SaaS provider I've used that has SSO support allows the customer to manage their IdP settings.

This is the scenario I am currently working with:
Our customer contacts do not work in IT and are not technical, in most cases their IT team is on another continent. In addition, many of our customers speak a different language than any of our employees and a translator is involved in all meetings.

We have to ask our customer contact to create an IT ticket in their internal helpdesk system asking their IT team to configure our application in their IdP and then pass along the information that we will need to configure the IdP in keycloak. Typically it takes weeks if not months to convince the customer contact to talk to their IT team. Once that happens (and assuming they actually pass on the correct information to their IT team post translation and telephone game) we then have to wait for their IT team to pick up the ticket and configure things on their end, which could take anywhere from weeks to a month or more. 

Once the IdP info we need it passed back to us, and assuming it's actually the information we need, we can then configure things on our end. Now we need to ask the customer contact to attempt to login, if it fails, we then have to get the customer contact to create an IT ticket that references the previous ticket and .... I hope you get the picture.

We need to be able to send an email that can be passed along to their IT team who then has all of the information they need to: 

1. configure our app in their IdP
2. login to our system
3. configure their IdP in our system
4. test that they are able to login without any additional back and forth

**Comment by sguilhen** [comment]
> By enforce I mean that when we move to 26 from 25 (without the experimental feature) will existing users get a "default org" assigned? Or is there something like a user without any org?

No, existing users won't be assigned to a default org. They will remain realm users and you can then onboard them to an org (or invite existing users to join the org).

**Comment by sguilhen** [comment]
> Will the organisations-feature support different (and/or extra) "User Profile > Attributes" per organisation? This is now a Realm setting.

For the moment it is still a realm-level setting but I can see the value of having org-specific attributes - but definitely not for Keycloak 26, most likely for a minor release after that if the priority is high enough or if we get it contributed by the community.

**Comment by sguilhen** [comment]
> I can see that a lot of effort was put into the orga feature which is great, but I could not find an outlook to when this core use-case will be covered so it is hard to actually plan using it. 26.1? 27 (in 2026?!)

@migu83 just to clarify, you're referring to the ability to assign roles to an org and have those available for the org members? If so, this is one of the things we would like to include but due to the limited bandwidth we can't deliver in time for 26. I personally don't think this would be too difficult to implement but it will have to wait for a minor release after 26.

**Comment by sguilhen** [comment]
> @pedroigor is it possible, that when I send someone an invitation to my organization and they register, that they are redirected to my web application? so basically the home url of my keycloak client

@Degoya-Studio I think this kinda of customization makes a lot of sense. I've filed https://github.com/keycloak/keycloak/issues/33201 to track this enhancement

**Comment by sguilhen** [comment]
> Is there any roadmap for when the additional features will be added? The organization self-service feature is a core capability that we need before we can use keycloak. Literally every SaaS provider I've used that has SSO support allows the customer to manage their IdP settings.
> 

I totally get the picture (and feel your pain, trust me). Self-service is one of the top features around orgs that we will be looking into for a future release. We understand it is an important point, but we're currently starting to review the fine-grained permissions we have today in Keycloak and the self-service organization console depends on that to some extent.

I hope we will be able to share some news around timelines soon.

**Comment by tomlnd** [comment]
Would the RBAC feature under the Administrative Authorization bullet point replace the need for this extension [https://github.com/sventorben/keycloak-restrict-client-auth](https://github.com/sventorben/keycloak-restrict-client-auth) when wanting to restrict which users can authenticate to specific clients while denying others?

E.g. would it be possible to link clients to organizations as well (or any other possible approach)?

**Comment by osajdapawel** [comment]
@pedroigor 
I created enhancement issue about ability of connecting IdP with more than one organization. I wanted to link it also here bcs as enhancement it can go without noticing. IMO it would be nice feature that would simplify realm management. 
#31705
[https://github.com/keycloak/keycloak/issues/31705](https://github.com/keycloak/keycloak/issues/31705)

**Comment by paulwer** [comment]
is there a roadmap for the outstanding features, when to exspect what?

**Comment by mikolajMacioszczyk** [comment]
> @pedroigor I created enhancement issue about ability of connecting IdP with more than one organization. I wanted to link it also here bcs as enhancement it can go without noticing. IMO it would be nice feature that would simplify realm management. #31705

I completely agree with @osajdapawel that it would be beneficial to have an IdProvider that supports multiple organizations. If the domain specified for the organization would match the user's e-mail domain, then the user will be assigned to the organization. If no such organization is found, the user could be added to a realm without any organization.

**Comment by SMichalet** [comment]
> > > Perhaps, is there a way to skip the identification phase if we know the organization up front?
> > 
> > 
> > +1 for the suggestion by @UXabre. I would also like to have an option to pre-select the organization for the browser authentication flow if I already know the organization in the application initiating the authentication flow (e.g., multi-tenant web app with the organization ID in the subdomain: .myapp.com).
> 
> @rspu Does it help #31438?
> 
> In regards to themes, if you try a nightly image you'll be able to access the organization that matched the user trying to authenticate from the `org` bean. So you can customize the pages accordingly to the current organization.

@pedroigor 
I tried the option to pre-select the organization for the browser with the scope "organization:org" but the browser still want to check the email domain before redirecting to the organization's IdP.
It shows the error in the form "Your email domain matches the org organization but you don't have an account yet" but I don't have enter any email address yet.

I don't find an other way to skip the identification if we already know the organization.

**Comment by Degoya-Studio** [comment]
Hi @pedroigor any updates on **Organization Self-Service**? it aligns with this task https://github.com/keycloak/keycloak/issues/31641 organization admin members should be able to invite other members, without having realm admin rights.

**Comment by Florzl** [comment]
> Hi @pedroigor any updates on **Organization Self-Service**? it aligns with this task #31641 organization admin members should be able to invite other members, without having realm admin rights.

So if you only want to invite other members, you can wrap the keycloak admin api into your own api, maybe set the label of the owner uuid in the organization labels, so that only he can invite people. Should do the work.

**Comment by Degoya-Studio** [comment]
> So if you only want to invite other members, you can wrap the keycloak admin api into your own api, maybe set the label of the owner uuid in the organization labels, so that only he can invite people. Should do the work.

Unfortunately it's not only that. An organization admin should also be able to remove users from the org and assign specific roles.

**Comment by Steve-Rithmik** [comment]
> > So if you only want to invite other members, you can wrap the keycloak admin api into your own api, maybe set the label of the owner uuid in the organization labels, so that only he can invite people. Should do the work.
> 
> Unfortunately it's not only that. An organization admin should also be able to remove users from the org and assign specific roles.

And manage their Org's SSO configuration.

**Comment by pedroigor** [comment]
Thanks all for the feedback so far. 

About https://github.com/keycloak/keycloak/issues/31705, we were also not sure about not allowing sharing brokers. We ended up with the constraint because it is easier to relax later. For me, it makes sense, and we are open to any contribution. Or work ourselves if we get enough feedback to mark it as important.

About self-service, we also see as a key feature for the whole CIAM given that it reduce administrative costs from realm administrators by delegating tasks to organization admins. We are about to start reviewing the capabilities we have around admin authorization (e.g.: rbac and fine-grained admin) and that will be a key capability to build the self-service. We have also been discussing it here https://github.com/keycloak/keycloak/discussions/34005#discussioncomment-10970191.

About the roadmap, we are prioritizing the backlog for organizations. We already have a good idea about what people are asking for:

* Self-Service
* Clients and Service Accounts
* Roles
* Authentication Policies
* SCIM

Not necessarily the order but that is what I have from the top of my head based on issues and discussions.

**Comment by pedroigor** [comment]
@SMichalet Thanks. The behavior you described is a bug. That message does not make sense at that step of the identity-first login flow. Created https://github.com/keycloak/keycloak/issues/34069.

**Comment by ggomarighetti** [comment]
@pedroigor Quite the opposite, thank you very much for this feature and all the hard work it entails.

> About #31705, we were also not sure about not allowing sharing brokers. We ended up with the constraint because it is easier to relax later. For me, it makes sense, and we are open to any contribution. Or work ourselves if we get enough feedback to mark it as important.

I recently encountered this issue while testing the newly released version. The current functionality requires repeatedly configuring the same authentication providers for each organization, which increases the administrative burden and, more importantly, the risk of inconsistencies. Sharing brokers centralizes this process and simplifies maintenance. For example, in companies with multiple divisions using the same external providers, sharing brokers reduces the administrative effort by avoiding duplicated configurations and promotes more agile management. Or in a multitenant SaaS application, where the organizations feature is used to separate clients while still sharing all features and configurations, such as social providers.

I’d also like to mention that it would be useful to implement (or document, as I couldn’t find any information) a process to prevent users who don’t belong to any organization from logging in. Something like a feature that allows, for a given client, login only if the user belongs to an organization within the realm, or the opposite: only if they are an individual not affiliated with any organization. Alternatively, it could maintain the current mixed approach, but allow for free choice.

**Comment by maxhov** [comment]
Has the Roles issue already been given some thought? Cannot seem to find a dedicated issue for it (or have I glanced over it?), so posting it here.  
How would the feature work in combination with client roles? Would you be able to assign a member a client role? This is more or less the only feature we are currently missing to adopt Organizations into our application and we are very keen to do so, it is great!

**Comment by ggomarighetti** [comment]
I also contribute to the discussion with the following issue, to enable filtering or conditioning user authentication to a specific organization. https://github.com/keycloak/keycloak/issues/34275

**Comment by Lord-KalEl** [comment]
At first, thanks for this required feature "organizations" … I wanted to leave KeyClok because of the lack of this future (the extension was an option, but this, I thought, has to be core in such a software.)

But now, the organizations are there, the first step is done. It makes less sense, to have only the organizations, without orga-roles or orga rules. This step is very important, because otherwise, the orgas are similar to the groups.
Next thing is the nested organizations … if you have many of them, you should be able to struct them. 
One of your features will be orga-groups.
Definition of a group in KeyCloak is, to group some users.
Definition of a orgas in KeyCloak is, to group some users, but to isolate the rules and polices (and theming and other features)

So the idea is, we need both, nested groups AND nested orgas.

By the way, themes should be changable without a restart of KeyCloak … and it would be nice, if every orga can have their own theme. (500 to 1000 organizations, the downtime will be very high ;) )

Greetings
Kal

**Comment by pedroigor** [comment]
@Lord-KalEl I agree 100% about :

```
It makes less sense, to have only the organizations, without orga-roles or orga rules. This step is very important, because otherwise, the orgas are similar to the groups.
```

As you mentioned, we delivered the core.

The three features you mention:

* Organization Groups
* Groups of Organizations
* Themes

They are in the feature roadmap as per this issue. Actually, except for Groups of Organizations, something we did not consider yet but definitely possible to achieve with the core stuff we delivered as we would be basically creating a top-level group for a set of organizations.

**Comment by Kobudzik** [comment]
> > I can see that a lot of effort was put into the orga feature which is great, but I could not find an outlook to when this core use-case will be covered so it is hard to actually plan using it. 26.1? 27 (in 2026?!)
> 
> @migu83 just to clarify, you're referring to the ability to assign roles to an org and have those available for the org members? If so, this is one of the things we would like to include but due to the limited bandwidth we can't deliver in time for 26. I personally don't think this would be too difficult to implement but it will have to wait for a minor release after 26.

Hello @sguilhen! What's the status of this feature? I see Keycloak is at version 26.0.5 right now, but I don't see related changes in the changelog. Is it still to be implemented?

Thanks in advance for the reply!

**Comment by 4FunAndProfit** [comment]
I agree with @Kobudzik, this feature [userA is member of orgA (has RoleX here) and orgB (has RoleY here)] is really what's missing for keycloak organization to be usable in real cases :)

Please @sguilhen make it happen soon if possible, it would be incredible! 😍😍

Thanks again to everyone for this great work, it's really great progress for an amazing tool ❤️

**Comment by SLedunois** [comment]
Do we have a roadmap for the organizations? I see a lot of improvements requested in the issues, but none planned? Organizations are a good first step, but I think people need more features like groups and roles like in the phasetwo extension

**Comment by oboudet-beetween** [comment]
> > I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.
> > I hope that in the future there will also be an "admin" console for administrators of individual organizations.
> 
> @aljaznuncic The "self-service" for organization administrations in among the capabilities planned for this feature. See the description of the issue. But we are not delivering it now in Keycloak 26.
> 
> In regard to the domain, does it work if you provide some random/dummy domain? Could you please elaborate more about your use case and if just the fact that the user is a member of the org is enough for you as well as being able to introspect tokens to check the organizations the user is a member?

Hello @pedroigor I have the same use case as @aljaznuncic : I will have some organizations with a linked IDP, so a domain will be associated to them. But I also will have organizations without IDP, and without domain as email can be @gmail.com or others public emails providers.
I would like to be able to create organization without domain, and be able to just add members manually in it, because my authentication flow will be based on organization attributes (we wrote a custom authenticator to handle our needs with organization attributes)

**Comment by olivierboudet** [comment]
> > I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.
> > I hope that in the future there will also be an "admin" console for administrators of individual organizations.
> 
> @aljaznuncic The "self-service" for organization administrations in among the capabilities planned for this feature. See the description of the issue. But we are not delivering it now in Keycloak 26.
> 
> In regard to the domain, does it work if you provide some random/dummy domain? Could you please elaborate more about your use case and if just the fact that the user is a member of the org is enough for you as well as being able to introspect tokens to check the organizations the user is a member?

Hello @pedroigor I have the same use case as @aljaznuncic : I will have some organizations with a linked IDP, so a domain will be associated to them. But I also will have organizations without IDP, and without domain as email can be @gmail.com or others public emails providers.
I would like to be able to create organization without domain, and be able to just add members manually in it, because my authentication flow will be based on organization attributes (we wrote a custom authenticator to handle our needs with organization attributes)

**Comment by Lord-KalEl** [comment]
> I can see that a lot of effort was put into the orga feature which is great, but I could not find an outlook to when this core use-case will be covered so it is hard to actually plan using it. 26.1? 27 (in 2026?!)

2026??? I hope, this will be earlier …
At least, some features should be there much earlier …

Greetings and good work,
Kal

**Comment by lam-lt** [comment]
Thanks for all the great work!

I have an idea that might be of interest to many organizations: How to force users (who are members of an organization linked to an Organization IDP) to log in via the Organization IDP and disallow logins via Keycloak.

**Details:**

- My company (for example, "X") has an Identity Provider (using OIDC).
- Keycloak serves as an Identity Broker. Keycloak creates an organization named "X" and links organization "X" to the Identity Provider.

After users are linked to the X IDP, they can currently log in to Keycloak in two ways:
- From the X IDP
- From the Keycloak login form

As a leader of my company, I want my users to log in only via the X IDP and prevent them from logging in (and resetting passwords) via the Keycloak login form.

Are there any existing ways to achieve this (that I may not be aware of)? If not, would you consider developing a solution based on this idea?

**Comment by vidya086** [comment]
Hi - I m wondering if we have something like organization level admin feature out of the box from release 26 . Or is this eomthing planned under - " Organization-specific Admin and User Events" for future release ? I have a use case where each organization should have a admin user which have access to manage organization settings and other users of that organization and at the same time it would mot have access to see other organization users. Any input on this is appreciated. Thanks.

**Comment by sguilhen** [comment]
The team is currently focused on improvements in the fine-grained admin permissions area, a feature that will be needed for self-service scenarios (in which org admins have the necessary permissions to manage their own orgs, including memberships, configuration of idps, etc).

However, this doesn't mean that improvements to orgs will be put aside indefinitely. The plan is to continuously deliver features in the next minor releases, and we will be prioritizing most of the tasks based on the feedback you guys have been providing here and in other discussions/issues. We already have a good sense of the "hot" areas - that is, the improvements that are being requested the most, and there are plans to draft a roadmap for the next steps early next year. We will share the plans here as soon as possible :)

**Comment by AustinOyugi** [comment]
I assumed you'd be using the https://github.com/p2-inc/keycloak-orgs extension, which has already solved the organization's issue fantastically, I've worked with it for about 2 years, and it seems okay!

**Comment by manoj-accelq** [comment]
Is organization specific password policy considered a valid feature? If so, is it in the roadmap?

**Comment by chmanoj133** [comment]
Is organization specific password policy and custom SMTP configuration considered as valid features? If so, is it in the roadmap?

**Comment by L-U-C-K-Y** [comment]
> I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.
> 
> I hope that in the future there will also be an "admin" console for administrators of individual organizations.

I fully agree, we have the exact same use-case where organization use the same @gmail domain:

Organization 1 - uses username + password
Organization 2 - uses @gmail workspace
Organization 3 - uses @gmail workspace

**Comment by CarlosRodrigues** [comment]
Hi everyone,

Is there anything planed in the roadmap, where role assignment can be done per organization?

Example: a user belonging to multiple tenants can have different role assignment, depending on tenant, while the roles themselves are realm-scoped.

A tree-like structure containing org-specific role assignments could then be added into the organizations claim, in the token, containing each org assigned roles for the user.


Thanks

**Comment by spodgorski** [comment]
Hi all,  I have two questions regarding organization and invitation email functionality.

1. Can I mark an email as a verified after user successfully clicks join organization? Right now I need to force 2 actions on user: joining the organization and also verify email. IMHO email should be marked as a verified when user clicked invitation email.

2. Can I mark user as Unmanaged for every organization? I have a scenario when user has been invited from the organization A then after some time also from organization B. Then organization A is deleted and the user should not be as it is part also of organization B. Right now together with the organization user is also deleted as it is marked as managed by this orgazanition

**Comment by mcgrawia** [comment]
I would like to second @spodgorski 's comment above. We are just about to launch a new multi-tenant service and are using Keycloak Organizations, and I came here to document I have the two exact same suggestions/concerns. We are able to work around 1 for now by disabling the email validation requirement, but it seems like clicking the invitation link should count as verification. 

The second point is more of a concern. We plan to let users be invited to multiple orgs, and users will only ever be added via organization sign up registration/invite emails, so users will always be tied to the first org they're invited to. This is going to be dangerous later when that org is deleted, they will be completely removed and need to be re-invited as a net new user with a new user ID to all orgs they had access to previously.

**Comment by spodgorski** [comment]
@mcgrawia We found a workaround for this situation. Before sending the invitation, we will create draft user account within the realm to ensure that they will not be managed by any organization. During the creation process via the Keycloak Admin API, we can set the email to be verified and enforce the "Update Password" action to require the user to set a password. Until then, the account cannot be used, as it does not have a password set.

After creation, we will generate an invitation again via the Keycloak Admin API and send the link to the user.

When the user completes the registration by clicking the link, we will have a fully created account that is not managed by any organization as it was initially created within the realm instead of by organization invitation and has a verified email. However, as I mentioned at the beginning, this is more of a workaround than an expected solution.

**Comment by mcgrawia** [comment]
thanks for the tip @spodgorski, that approach does work, but produces a cumbersome registration flow for new users. It involves a total of 5 users actions before a user gets into our app:

1. User clicks the link in their email
2. Users first see a confirmation screen to "join" the organization, which is strange because the organization was created as part of their account set up, so they shouldn't need to confirm to join their own org
3. Users then need to set a password
4. After setting the password, users get a "Your account has been updated" screen they also have to click through
5. Lastly, the user is asked to sign in again after they just entered their password on the previous screen

All the above steps are required just to get the new user into the app, then they have to go through our app's account set up pages next, so we're trying to minimize the amount of pages/clicks users need to make to register. 

The nice part about the direct org-invite + registration flow is it's a much simpler user experience, with only two steps:

1. User clicks the link in the email
2. User enters their information and password and is logged into the app

Of course, this has the downside users become tied to their first organization's lifecycle.

**Comment by sguilhen** [comment]
> Is organization specific password policy and custom SMTP configuration considered as valid features? If so, is it in the roadmap?

Absolutely, it makes sense. The team will have a meeting shortly to define goals/priorities for the upcoming releases, and working on the next batch of organizations functionalities is definitely in the list. As such, all comments/votes here will be taken into consideration when we scope the work for the next releases.

**Comment by sguilhen** [comment]
> > I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.
> > I hope that in the future there will also be an "admin" console for administrators of individual organizations.
> 
> I fully agree, we have the exact same use-case where organization use the same [@gmail](https://github.com/gmail) domain:
> 
> Organization 1 - uses username + password Organization 2 - uses [@gmail](https://github.com/gmail) workspace Organization 3 - uses [@gmail](https://github.com/gmail) workspace

Noted - will take this to the team for further discussions.

**Comment by sguilhen** [comment]
> Hi everyone,
> 
> Is there anything planed in the roadmap, where role assignment can be done per organization?
> 
> Example: a user belonging to multiple tenants can have different role assignment, depending on tenant, while the roles themselves are realm-scoped.
> 
> A tree-like structure containing org-specific role assignments could then be added into the organizations claim, in the token, containing each org assigned roles for the user.
> 
> Thanks

Hi @aljaznuncic  - yes, this is in the roadmap. As I mentioned in a previous post, the team will be meeting later this month and then we will be able to share what the priorities for the next releases will be in terms of org features.

organization-specific  roles and groups are two of the most requested features and will likely be on the list.

**Comment by sguilhen** [comment]
> Hi all,
> 
> This is a highly requested feature and its great to see some progress here! A few years back we implemented our own organization management on top of keycloak and we are looking forward to switching to a native solution.
> 
> Are there any details for the individual capabilities provided by Keycloak Organizations in the list above? For example what does "Organization Self-Service" mean specifically?
> 
> Also is there a high-level roadmap on when we can expect new capabilities added? The milestones unfortunately do not contain any issues with label `area/organizations` yet.
> 
> Thanks a lot for your insights and work on this great feature!

Yeah, we need to organize the issues/milestones a bit to make it clearer what capabilities already exist and what is yet to be added.

When we talk about self-service, it is the ability to give certain users the role of "org-admin" or something like that, allowing this user to manage the org without necessarily having realm-level permissions, which are often too broad. Think of it as a stripped-down version of the admin console, allowing the org admins to manage their own members, idps, smtp settings (yet to be added), etc.

**Comment by sguilhen** [comment]
> KC team: I'm not sure if it would be possible to add a boolean flag to the organization invite API that would allow our BE to set the membership type on the new user? If we could tell Keycloak to create the user as unmanaged, I think this would work as we need.

We will look into it :)

**Comment by pedroigor** [comment]
> > Hi all,
> > This is a highly requested feature and its great to see some progress here! A few years back we implemented our own organization management on top of keycloak and we are looking forward to switching to a native solution.
> > Are there any details for the individual capabilities provided by Keycloak Organizations in the list above? For example what does "Organization Self-Service" mean specifically?
> > Also is there a high-level roadmap on when we can expect new capabilities added? The milestones unfortunately do not contain any issues with label `area/organizations` yet.
> > Thanks a lot for your insights and work on this great feature!
> 
> Yeah, we need to organize the issues/milestones a bit to make it clearer what capabilities already exist and what is yet to be added.
> 

This issue already holds all the existing capabilities (they are checked), and those still missing. What we don't have yet is any planned work to deliver the remaining capabilities which then will be added (and updated here) with their respective milestones. This is kinda how we run the first iteration.

**Comment by pedroigor** [comment]
@sguilhen I can already think about two issues we can include in this issue/roadmap, but I'm sure there are others from some other comments:

* Support for SMTP configuration
* Support for selecting the membership type when sending invitation links

The SMTP configuration should be an item in this issue as we don't yet have the capability. But for the second one we should create an `enhancement` and have it in our backlog given that it is related to an existing capability.

Eventually, we should create issues for each item listed in this issue/roadmap so that people can vote for/ask for a specific capability and help us decide which ones to prioritize.

**Comment by paulwer** [comment]
> > > I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.
> > > I hope that in the future there will also be an "admin" console for administrators of individual organizations.
> > 
> > I fully agree, we have the exact same use-case where organization use the same [@gmail](https://github.com/gmail) domain:
> > 
> > Organization 1 - uses username + password Organization 2 - uses [@gmail](https://github.com/gmail) workspace Organization 3 - uses [@gmail](https://github.com/gmail) workspace
> 
> Noted - will take this to the team for further discussions.

Hi @sguilhen ,
there is an issue with a very good suggestion regarding handling the domain and login flow related to it.
https://github.com/keycloak/keycloak/issues/31285
Happy if you considder this within your discussion.

Thanks yall :)

**Comment by sguilhen** [comment]
> > > > I am happy for a good feature. However, it would be very welcome if the domain was not required when creating the organization. In our application, we have companies that use @gmail.com or the internet provider's domain for their email, or others that have a domain at the level of the entire corporation and would like to have their own organization within our application (they do not have a subdomain). For the last example, a checkbox would be welcome whether you really want users to be automatically added to the organization based on the given domain or not.
> > > > I hope that in the future there will also be an "admin" console for administrators of individual organizations.
> > > 
> > > 
> > > I fully agree, we have the exact same use-case where organization use the same [@gmail](https://github.com/gmail) domain:
> > > Organization 1 - uses username + password Organization 2 - uses [@gmail](https://github.com/gmail) workspace Organization 3 - uses [@gmail](https://github.com/gmail) workspace
> > 
> > 
> > Noted - will take this to the team for further discussions.
> 
> Hi [@sguilhen](https://github.com/sguilhen) , there is an issue with a very good suggestion regarding handling the domain and login flow related to it. [#31285](https://github.com/keycloak/keycloak/issues/31285) Happy if you considder this within your discussion.
> 
> Thanks yall :)

Hi @paulwer , thanks for the pointer. @pedroigor Do you think we should link it here or do you have a better option to organize the enhancements that are not exactly new capabilities?

**Comment by fosple** [comment]
Thanks for the update @sguilhen & @pedroigor ❤ Already looking forward to more organization features and happy to beta test 🚀

**Comment by wingsofovnia** [comment]
Seems not to be mentioned yet, another use case for organisations is using it for SaaS multitenancy with KC alone. In my project, I created a required action to enforce organization creation for new user registrations, unless they are invited by a link. This way every user is related to an org (tenant) and I can authorise their requests accordingly. Here, having domains has no sense and I just set random value in the required action.

**Comment by Gianpiero-3di** [comment]
Correct me if I'm wrong:
currently isn't possible to have the same user granted different permission sets on an organization basis, right?
I mean, a situation where an access token would expose something like:
```json
{
...
  "organizations": {
    "permissions": {
      "Org_123": ["doc:edit", "doc:view"],
      "Org_456": ["doc:view", "post:create", "post:edit"]
    }
  }
}
```
And is it supposed to be added in the future?

**Comment by wingsofovnia** [comment]
> I created a required action to enforce organization creation for new user registrations - [#issuecomment-2764329298](https://github.com/keycloak/keycloak/issues/30180#issuecomment-2764329298)

If anyone is interested, posted under MIT: https://github.com/wingsofovnia/keycloak-organization-required-action

**Comment by fosple** [comment]
Two suggestions for features regarding the user invitation process for organizations:

1. When inviting an user to an organization, would be nice to also attach a role to this invitation. So when the user signs up she is already assigned to that role.

Example from notion.com:

<img width="350" alt="Image" src="https://github.com/user-attachments/assets/51d86f51-3f36-4b87-b61c-df608cd1ad0b" />

2. When inviting an user to an organization, also allow the option for the user to just register via Keycloak, performing the e-mail validation, and then accept or reject the invitation to the organization. This means no special invite link is needed. 

Example from notion.com: screen directly after registering and validating the email via a code I got via mail (workspace is here equivalent to organization):

<img width="350" alt="Image" src="https://github.com/user-attachments/assets/4c8324ef-c754-4875-baca-02932bf9517c" />

(In general I'm also looking forward to the "Manage Invitations" part of the Organizations feature :) )

**Comment by SferaDev** [comment]
Similar to @fosple we are also using Organizations for handling multi tenancy in our SaaS and we're missing Organization Membership Roles and RBAC features.

**Comment by HugoMuller** [comment]
Hi,

> Support for selecting the membership type when sending invitation links

Did you plan to also support membership update in the Admin UI, or with the admin API?

As some of my clients have requested SSO, I have to plug Keycloak to my app. So I have migrated all the users and organizations from my DB to Keycloak. The thing is that all users belonging to an organization are now "unmanaged".
However, I need to set "managed" to all members of a particular organization, but this seems not possible at the moment.

Thanks for this great feature!

**Comment by oculos** [comment]
I was wondering something:
Organizations might be very interesting for us in the sense of separation of internal and external users. 
The problem is the reliance on a domain name for organization membership.
In your documentation (https://www.keycloak.org/docs/latest/server_admin/index.html#creating-an-organization), a Domain name doesn't seem to be required (see picture there). 
Experimenting with organizations, I see that domain names are indeed required to setup an organization. 

Couldn't one simply rely on the Identity Provider for defining organization membership?

**Comment by SferaDev** [comment]
@oculos I have this PR in review for that: https://github.com/keycloak/keycloak/pull/40508

**Comment by oculos** [comment]
> [@oculos](https://github.com/oculos) I have this PR in review for that: [#40508](https://github.com/keycloak/keycloak/pull/40508)

This is awesome! I really hope it goes through!

**Comment by spodgorski** [comment]
"@SferaDev Do you have an estimated release date for optional domains? We're about to go live, and we're really counting on this feature — our customers might not have their own domains, but we still need the organization functionality to manage their members."

**Comment by SferaDev** [comment]
@spodgorski Unfortunately no, we're also waiting on the feature. For now we're adding invalid domain strings (like an uuid without tld).

Both our contributions for [making domains optional](https://github.com/keycloak/keycloak/pull/40508) and [managing organization invites](https://github.com/keycloak/keycloak/pull/40526) are waiting for @pedroigor review.

**Comment by FreeSynergy** [comment]
In my case, I'm waiting for the "nested" and the self organized organizations … I need an admin role or alternative roles for organizations … (Maybe a top-down or/and a bottom-up inheritance)

I will make in the next days a new issue out of that - with much more ideas …

Thanks …

**Comment by paulwer** [comment]
@pedroigor are there any timeline updates regarding the features planned or for the community PRs?

**Comment by pedroigor** [comment]
> [@pedroigor](https://github.com/pedroigor) are there any timeline updates regarding the features planned or for the community PRs?

I know we are not keeping pace with the demands and planned capabilities for this feature. There is no timeline update, even though we were considering adding support for roles in 27. However, we are running to deliver a new feature while at the same time keeping pace with issues from the community (bugs and high-priority enhancements).

That is why we have not given any updates since we released the feature. We have made only a few small improvements, but nothing yet that significantly impacts the capabilities herein defined.

**Comment by fosple** [comment]
@pedroigor Thanks for the update. When you come back to this feature and need inspiration definitely have a look at [authkit.com](https://authkit.com) by [workos.com](https://workos.com) - it's used by OpenAi, Perplexity, Cursor,... and solves Organizations-Requirements really well. For now I will use them until we'll have a real Open Source alternative. Thanks for all the effort you guys put into Keycloak - much appreciated. Hope I can use Keycloak-Organizations for my next startup then ;)

**Comment by jtejido** [comment]
It would be nice if we can get the OrganizationModel from UserSessionModel.getUser(). It means we can utilize it and reach multi-tenant URLs to enrich our token via [custom mappers](https://www.keycloak.org/docs-api/latest/javadocs/org/keycloak/protocol/oidc/mappers/AbstractOIDCProtocolMapper.html#setClaim(org.keycloak.representations.IDToken,org.keycloak.models.ProtocolMapperModel,org.keycloak.models.UserSessionModel,org.keycloak.models.KeycloakSession,org.keycloak.models.ClientSessionContext)) , where hosts are in the form of \<org>.acme.com/claims

**Comment by rasos** [comment]
The [release notes for 26.5.0](https://github.com/keycloak/keycloak/releases/tag/26.5.0) mention some great advancements with invitation management, starting with the sentence  

> Organization administrators can now manage organization invitations through both the Admin Console and REST API

However we could not find such a (new) role to manage an organization https://github.com/keycloak/keycloak/issues/31641 yet. Still, only a realm administrator can manage organizations. I guess this comes when introducing roles https://github.com/keycloak/keycloak/issues/40585 .

**Comment by paulwer** [comment]
@pedroigor is it planned / would it be possible to add organization-workflow integration to this epic as well?
(or is this tracked somewhere else?)

I think of following use cases:
- apply admin role to first user joining an organization / apply role to last remaining user in organization etc.
- remove org when containing no members and org is stale for defined time-period
- welcome message for joining users
- manage user-groups/-attributes based on organization memberships

**Comment by Kalffman** [comment]
@paulwer Just adding my two cents here.

It might be worth exploring whether we can listen to subscription events for the org groups and mirror those to the realm group subscriptions.

My thinking is that this could get us a solid first step toward B2B role management—kind of a Pareto-style win, where relatively small effort gets us most of the way there.

**Comment by ricardomarks** [comment]
Hey guys, I'm using the Organization feature and we're excited about its future in Keycloak. 

My question/doubt would be whether in the future, it will be possible to login, through the organization:alias scope, to correctly filter the groups and roles referring to the org in the JWT token, in addition to blocking the login when the user does not belong to the requested org. Here I'm talking about client-roles, realm-roles, groups-membership mappers.

**Comment by vramik** [comment]
> Hey guys, I'm using the Organization feature and we're excited about its future in Keycloak.
> 
> My question/doubt would be whether in the future, it will be possible to login, through the organization:alias scope, to correctly filter the groups and roles referring to the org in the JWT token, in addition to blocking the login when the user does not belong to the requested org. Here I'm talking about client-roles, realm-roles, groups-membership mappers.

  Hi @ricardomarks, most of what you're describing is already supported today:
  
  Login blocking — when you request organization:<alias> scope and the user is not a member of that organization, the request is rejected. This also applies on token refresh — if membership is revoked after login, the refresh will fail.

  Groups — realm groups and organization groups are separated in the token. The standard group membership mapper only includes realm groups. Organization groups appear nested under the organization claim, scoped to each organization. So when you log in with organization:<alias>, you already get the org-filtered group view.

  Roles (available in `nightly` and since `26.7.0`) — the top-level `realm_access` and `resource_access` claims always include all user roles globally, regardless of which organization was requested. But, if you assign roles through organization groups, you can enable the `addGroupRoleMappings` option on the organization group membership mapper, which will nest those roles under the organization claim. But there's no mechanism (at least today) to filter the top-level role claims by organization context.

**Comment by paulwer** [comment]
Is it planned to support organization profile management, like it is implemented for users already?
Would that be a seperate point on the list above or would that already be part of self-service?
